### ISSUE TITLE
Quantization issue in transformers

### ISSUE THREAD

**Issue by linbinyu** [pytorch/pytorch#82443]
### 🐛 Describe the bug

This issue happens in when quantization a simple transformer model (1.13)

Example
`
    class M(torch.nn.Module):
        def __init__(self):
            super(DynamicQuantModule.M, self).__init__()
            self.transformer = nn.Transformer(d_model=2, nhead=2, num_encoder_layers=1, num_decoder_layers=1)

        def forward(self):
            return self.transformer(torch.randn(1, 16, 2))
    
torch.quantization.quantize_dynamic(M(), dtype=torch.qint8)
`
The error is
`
  File "/Users/linbin/opt/anaconda3/lib/python3.8/site-packages/torch/jit/_recursive.py", line 516, in init_fn
    scripted = create_script_module_impl(orig_value, sub_concrete_type, stubs_fn)
  File "/Users/linbin/opt/anaconda3/lib/python3.8/site-packages/torch/jit/_recursive.py", line 542, in create_script_module_impl
    create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
  File "/Users/linbin/opt/anaconda3/lib/python3.8/site-packages/torch/jit/_recursive.py", line 393, in create_methods_and_properties_from_stubs
    concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
RuntimeError:
method cannot be used as a value:
  File "/Users/linbin/opt/anaconda3/lib/python3.8/site-packages/torch/nn/modules/transformer.py", line 468
                self.norm2.weight,
                self.norm2.bias,
                self.linear1.weight,
                ~~~~~~~~~~~~~~~~~~~ <--- HERE
                self.linear1.bias,
                self.linear2.weight,
`


### Versions

PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (x86_64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.18.6)
CMake version: version 3.21.3
Libc version: N/A

Python version: 3.7.5 (default, Oct 22 2019, 10:35:10)  [Clang 10.0.1 (clang-1001.0.46.4)] (64-bit runtime)
Python platform: Darwin-21.5.0-x86_64-i386-64bit
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] No relevant packages
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.2.0           hecd8cb5_269  
[conda] mkl-include               2022.0.0           hecd8cb5_105  
[conda] mkl-service               2.4.0            py38h9ed2024_0  
[conda] mkl_fft                   1.3.0            py38h4a7008c_2  
[conda] mkl_random                1.2.2            py38hb2f4e1b_0  
[conda] numpy                     1.20.0                   pypi_0    pypi
[conda] numpy-base                1.20.2           py38he0bd621_0  
[conda] numpydoc                  1.4.0            py38hecd8cb5_0  
[conda] pytorch                   1.13.0.dev20220728         py3.8_0    pytorch-nightly
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torchaudio                0.13.0.dev20220728        py38_cpu    pytorch-nightly
[conda] torchvision               0.13.0                   pypi_0    pypi

**Comment by kit1980** [comment]
The PR that introduced this likely landed Jul 15 2022, judging from when iOS tests started to fail (it's not clear when exactly, because the test uses nightly build to generate models).

This issue currently blocking enabling DynamicQuantModule in iOS simulator tests: https://github.com/pytorch/pytorch/issues/82139 

cc @pytorch/pytorch-dev-infra

**Comment by linbinyu** [comment]
Another way to trigger it is run:
```
python3 test/mobile/model_test/gen_test_model.py dynamic_quant_ops
```
in the latest nightly build.

**Comment by linbinyu** [comment]
https://github.com/pytorch/pytorch/pull/81277 was landed on July 15

**Comment by z-a-f** [comment]
Oooh, I don't think we can quantize that module in eager mode. Once quantized, the `.weight` becomes a method for "unpacking" the quantized weights. I think the only way to quantize this type of layers is by using the "fx quantization".

cc @jerryzh168

**Comment by jerryzh168** [comment]
yeah eager mode quantization does not support quantizing a linear that is used in this way (direct access to weight), is fx graph mode quantization an option here?

**Comment by linbinyu** [comment]
yes it works before. What is the right way?

**Comment by jerryzh168** [comment]
I'm wondering what changed? we didn't touch eager mode quantization code for a long time I think. Maybe one thing you can do is to add `@property` for https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/linear.py#L229 not sure if it would work or not though

**Comment by kit1980** [comment]
Any updates on this?
DynamicQuantModule in iOS simulator tests is still not running, it's better to be able to re-enable it soon to prevent regressions.

**Comment by jerryzh168** [comment]
I can't repro the issue actually, are you using pytorch master?

**Comment by linbinyu** [comment]
maybe not the root cause, but the crash site was touched by https://github.com/pytorch/pytorch/pull/81277 and https://github.com/pytorch/pytorch/pull/81013. 

To repro just run this command on the latest nightly:
python3 test/mobile/model_test/gen_test_model.py dynamic_quant_ops

If it works, then we can just enable the ios test.

**Comment by kit1980** [comment]
I've just tried rebasing https://github.com/pytorch/pytorch/pull/82439 and the issue is still there.

**Comment by jerryzh168** [comment]
<img width="1177" alt="Screen Shot 2022-08-15 at 5 36 21 PM" src="https://user-images.githubusercontent.com/4958441/184759099-c5d6b33e-bc86-4b4c-9c91-185793e32ba8.png">
is this the code to repro the issue?

**Comment by jerryzh168** [comment]
I see, it failed during torch.jit.script, basically after we quantize the model, linear.weight became a method instead of a value, that's why torch.jit.script failed. I really doubt that this worked before actually

**Comment by kit1980** [comment]
> I see, it failed during torch.jit.script, basically after we quantize the model, linear.weight became a method instead of a value, that's why torch.jit.script failed. I really doubt that this worked before actually

It worked before Jul 16th 2022.
See this PR https://github.com/pytorch/pytorch/pull/82027
Before Jul 16th DynamicQuantModule in the test worked, after that it needed to be disabled - otherwise it fails with the error from the first post in this issue.

**Comment by jerryzh168** [comment]
> > I see, it failed during torch.jit.script, basically after we quantize the model, linear.weight became a method instead of a value, that's why torch.jit.script failed. I really doubt that this worked before actually
> 
> It worked before Jul 16th 2022. See this PR #82027 Before Jul 16th DynamicQuantModule in the test worked, after that it needed to be disabled - otherwise it fails with the error from the first post in this issue.

is it possible to trace to the PR when this PR is broken?

**Comment by jerryzh168** [comment]
I don't have any updates, we haven't found the root cause of this issue I think

**Comment by jhoareau** [comment]
Hi @linbinyu @jerryzh168 - I am experiencing this issue when trying to quantize a PyTorch Transformer instance. I need to use ONNX Runtime instead, but it comes with its own drawbacks. We'd love to stick to pure PyTorch.
Is anyone looking at this issue?

**Comment by voladoddi** [comment]
Hi folks - wondering if anyone successfully was able to quantize a vanilla Transformer model or a `nn.Transformer` layer within PyTorch? I face the same issue.
